File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        int n = nums1.size(), m = nums2.size();

        stack<int> s;
        vector<int> temp(m);
        unordered_map<int, int> mp;

        for(int i = m - 1; i >= 0; i--){
            int val = nums2[i];
            while(!s.empty() && val >= s.top()){
                s.pop();
            }

            if(s.empty()){
                temp[i] = -1;
            } else{
                temp[i] = s.top();
            }

            s.push(val);
            mp[val] = i;
        }

        vector<int> ans;
        ans.reserve(n);

        for(int i : nums1){
            ans.push_back(temp[mp[i]]);
        }

        return ans;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int n1 = nums1.length, n2 = nums2.length;

        Stack<Integer> s = new Stack<>();
        HashMap<Integer, Integer> mp = new HashMap<>();

        for(int i = n2 - 1; i >= 0; i--){
            while(!s.isEmpty() && s.peek() <= nums2[i]){
                s.pop();
            }
            mp.put(nums2[i], (s.isEmpty() ? -1 : s.peek()));
            s.add(nums2[i]);
        }

        int ans[] = new int[n1];
        for(int i = 0; i < n1; i++){
            ans[i] = mp.get(nums1[i]);
        }
        
        return ans;
    }
}